docs/adr/0003-swift-packaging.md

```diff
@@ -0,0 +1,225 @@
+* Consistency with other iOS components being developed at Mozilla.
+* Ability for the Nimbus Swift bindings to easily depend on Glean.
+* Ease of maintainability for application-services developers.
+
+## Considered Options
+
+* **(A) Do Nothing**
+  * Keep our current build and distribution setup as-is.
+* **(B) Use Carthage to build XCFramework bundles**
+  * Make a minimal change to our Carthage setup so that it builds
+    the newer XCFramework format, which can support M1 Apple Silicon.
+* **(C) Distribute a single pre-compiled Swift Package**
+  * Convert the all-in-one `MozillaAppServices` Carthage build to a similar
+    all-in-one Swift Package, distributed as a binary artifact.
+* **(D) Distribute multiple source-based Swift Package targets, with pre-compiled Rust code**
+  * Split the all-in-one `MozillaAppServices` Carthage build into a separate
+    Swift Package target for each component, with a shared dependency on pre-compiled
+    Rust code as a binary artiact.
+
+## Decision Outcome
+
+Chosen option: **(D) Distribute multiple source-based Swift Packages, with pre-compiled Rust code**.
+
+This option will provide the best long-term consumer experience for iOS developers, and
+has the potential to simplify maintenance for application-services developers after an
+initial investment of effort.
+
+### Positive Consequences
+
+* Swift packages are very convenient to consume in newer versions of XCode.
+* Different iOS apps can choose to import a different subset of the available components,
+  potentiallying helping keep application size down.
+* Avoids issues with mis-matched Swift version between application-services build
+  and consumers, since Swift files are distributed in source form.
+* Encourages better conceptual separation between Swift code for different components;
+  e.g. it will make it possible for two Swift components to define an item of the same
+  name without conflicts.
+* Reduces the need to use XCode as part of application-services build process, in favour
+  of command-line tools.
+
+### Negative Consequences
+
+* More up-front work to move to this new setup.
+* May be harder to concurrently publish a Carthage framework for current consumers who aren't
+  able to move to Swift packages.
+* There is likely to be some amount of API breakage for existing consumers, if only in having
+  to replace a single `import MozillaAppServices` with independent imports of each component.
+
+### Implementation Sketch
+
+We will maintain the existing Carthage build infrastructure in the application-services repo and continue publishing a pre-built Carthage framework, to support firefox-ios until they migrate to Swift Packages.
+
+We will add an additional iOS build task in the application-services repo, that builds *just the Rust code* as a `.xcframework` bundle. An initial prototype shows that this can be achieved using a relatively straightforward shell script, rather than requiring a second XCode project. It will be
```

Comment:
(very naive thought incoming): when we talk about the `Rust Code` here, we mean _all_ the rust code right? like, the fxa_client component's rust code will still include the nimbus's rust code?

If not, ignore me 😬 - if it does include all the code, is there a world where each component only has a separate chunk of the Rust code that it needs? as I'm writing this, I imagine there's an easy answer to why it's `no`, but wanted to double check

Reply:
Yes, here "the Rust code" means "the Rust code for all components, compiled together into a single big blob". We expect this will actually be OK in practice because we're building it as a staticlib, so when it gets incorporated into the final application build, the compiler will only pull in the parts of that lib that it needs.

We've had quite a few discussions about being able to distribute different components with separately-compiled chunks of the Rust code, but it's tricky because the components sometimes want to talk to each other directly, and in order to do that they need to be compiled into the same artifact.

I have an imaginary world in my head where the components can each be distributed as little `.so` files that call each other through their C FFI layer, but that world is a long way from being reality and I don't think we have any concrete plans to head in that direction.